pkg/apis/garden/types.go

```diff
@@ -1302,6 +1302,106 @@ type KubeletConfig struct {
+	//   nodefs.available:   "5%"
+	//   nodefs.inodesFree:  "5%"
+	//   imagefs.available:  "5%"
+	//   imagefs.inodesFree: "5%"
+	EvictionHard *KubeletConfigEviction
+	// EvictionSoft describes a set of eviction thresholds (e.g. memory.available<1.5Gi) that if met over a corresponding grace period would trigger a Pod eviction.
+	// +optional
+	// Default:
+	//   memory.available:   "200Mi/1.5Gi/10%"
+	//   nodefs.available:   "10%"
+	//   nodefs.inodesFree:  "10%"
+	//   imagefs.available:  "10%"
+	//   imagefs.inodesFree: "10%"
+	EvictionSoft *KubeletConfigEviction
+	// EvictionSoftGracePeriod describes a set of eviction grace periods (e.g. memory.available=1m30s) that correspond to how long a soft eviction threshold must hold before triggering a Pod eviction.
+	// +optional
+	// Default:
+	//   memory.available:   1m30s
+	//   nodefs.available:   1m30s
+	//   nodefs.inodesFree:  1m30s
+	//   imagefs.available:  1m30s
+	//   imagefs.inodesFree: 1m30s
+	EvictionSoftGracePeriod *KubeletConfigEvictionSoftGracePeriod
+	// EvictionMinimumReclaim configures the amount of resources below the configured eviction threshold that the kubelet attempts to reclaim whenever the kubelet observes resource pressure.
+	// +optional
+	// Default: 0 for each resource
+	EvictionMinimumReclaim *KubeletConfigEvictionMinimumReclaim
+	// EvictionPressureTransitionPeriod is the duration for which the kubelet has to wait before transitioning out of an eviction pressure condition.
+	// +optional
+	// Default: 4m0s
+	EvictionPressureTransitionPeriod *metav1.Duration
+	// EvictionMaxPodGracePeriod describes the maximum allowed grace period (in seconds) to use when terminating pods in response to a soft eviction threshold being met.
+	// +optional
+	// Default: 90
+	EvictionMaxPodGracePeriod *int32
+}
+
+// KubeletConfigEviction contains kubelet eviction thresholds supporting either a resource.Quantity or a percentage based value.
+type KubeletConfigEviction struct {
+	// MemoryAvailable is the threshold for the free memory on the host server.
+	// +optional
+	MemoryAvailable *string
+	// ImagefsAvailable is the threshold for the free disk space in the imagefs filesystem (docker images and container writable layers).
+	// +optional
+	ImagefsAvailable *string
+	// ImagefsInodesFree is the threshold for the available inodes in the imagefs filesystem.
+	// +optional
+	ImagefsInodesFree *string
+	// NodefsAvailable is the threshold for the free disk space in the nodefs filesystem (docker volumes, logs, etc).
+	// +optional
+	NodefsAvailable *string
+	// NodefsInodesFree is the threshold for the available inodes in the nodefs filesystem.
+	// +optional
+	NodefsInodesFree *string
+}
+
+// KubeletConfigEviction contains configuration for the kubelet eviction minimum reclaim.
+type KubeletConfigEvictionMinimumReclaim struct {
+	// MemoryAvailable is the threshold for the memory reclaim on the host server.
+	// +optional
+	MemoryAvailable *resource.Quantity
```

Comment:
I think it makes sense to create a separate struct with these five properties and embed it then everywhere. It's not needed to define it three times.

Reply:
The reason I did it this way is: we know that the KubeletMinimumReclaim can only be a resource.Quantity (not a percentage). Also the SoftGracePeriod can only be a duration. The HardEviction and SoftEviction can either be a percentage or resource.Quantity, so I chose to most generic data type for it. 

If you want I can change it to one struct having the more generic datatype string and do the parsing in the storage validation.

Reply:
Hm, okay, I see. Valid argument, then let's keep it as it is.